### DISCUSSION TITLE
routing

### DISCUSSION THREAD

**Discussion by joamatab** [gdsfactory/gdsfactory#424] [General]
thank you  @thomasdorch  for your latest PR where we can use `get_route` for path length matching

See documentation in the latest 5.8.3 release
https://gdsfactory.github.io/gdsfactory/notebooks/04_routing.html#get_bundle_path_length_match

what was the motivation for not using get_route_path_length_match?

**Comment by thomasdorch** [comment]
Hi @joamatab. The issue I was having is that routing using `get_bundle_path_length_match` wouldn't find routes for more complicated bundles because it wasn't doing the checks that `get_bundle` does. Initially I made some changes so that `get_bundle_path_length_match` could take the `_get_bundle_udirect_waypoints` and `_get_bundle_uindirect_waypoints` and use those instead of `_get_bundle_waypoints` if needed. As I used it though it felt more intuitive to just call `get_bundle` when I want to route a bundle, and use kwargs to select the type of bundle (automatic, steps, or waypoints) and whether it should have path length matching or not, since the alternative was duplicating the functionality of `get_bundle` within `get_bundle_path_length_match`. 

Does this make sense?

**Comment by joamatab** [comment]
yes, it does, thank you for the clarification, `get_bundle` is more powerful so it makes sense to adapt it for path length matching as you did

what do you think of deleting the old `get_bundle_path_length_match`?

@tvt173

**Comment by tvt173** [comment]
ah cool, i'm not a big user of the path length matching, but i'm definitely more inclined to if it's a part of `get_bundle()`.

i have no strong opinions on deleting the old one. but if it's redundant, i'm fine with simplifying things

it would be nice in the future to allow supplying a list of desired delays for each path

**Reply by tvt173** [reply]
i also feel like it might be cleaner to have `pathlength_matching` be the argument to get bundle that takes a dict or basemodel of sub-settings for the pathlength matching. 

i.e.

```python
get_bundle(ports1, ports2,
                    pathlength_matching= dict(
                        match_loops=2,
                        match_extra_length=15,
                        ),
                    )
```

would look quite a bit cleaner in yaml too, i think

**Comment by thomasdorch** [comment]
I really like the idea of passing the length matching kwargs as a dict. Maybe we can take that further and make a BaseModel called RouterSettings or something like that and keep all of the routing parameters in it? Since the routing function settings aren't defined in the signature for `get_bundle`, this might make it easier to keep track of them.

**Comment by thomasdorch** [comment]
Another thing I've been considering is whether the default cross_section parameter should be set to None, and instead have the routing function get the cross section from the port cross section if they're all the same, and raise an error asking the user to specify a cross section otherwise. Additionally, we could pass in a RouterSettings BaseSettings that has both electrical and optical parameters and have the routing function get the route type (electrical or optical) from the port if the start and end ports have the same type.

Essentially like:
```
class RouterSettings(BaseModel):
    width: float | str = 'auto'
    cross_section: CrossSectionSpec = 'auto'
    bend: ComponentSpec = 'auto'
    straight: ComponentSpec = 'auto'
    taper: ComponentSpec = None
    path_length_matching: dict = dict(match_loops=1, match_extra_length=0)
    
class Router(BaseSettings):
    optical_settings: RouterSettings = RouterSettings()
    electrical_settings: RouterSettings = RouterSettings()


router = Router(
    optical_settings = RouterSettings(
        width=0.5,
        cross_section='strip'
    )
)
bundle = get_route(ports1, ports2, router=router)
```

where 'auto' means to pull it from the port

I'm curious what you think of this, as it would require a decent refactor but would also consolidate a lot of routing functionality and put the ports in control over the routes between them (unless overridden), since they already have some of this info. 

@joamatab @tvt173

**Reply by tvt173** [reply]
i think this is something that's maybe a bit dependent on the application and preference of the organization, but probably makes sense to store as a setting in the PDK.

what I find to be more common/convenient is to have a standard route type, always use that by default, and auto-taper to it as necessary. i have some code to do auto-tapering... it's not too hard to implement. but to do it well, we also need to register default transitions in the PDK. if the router can find a suitable transition between the route's cross section type and the port, it auto-transitions. if not, it throws an error. 

would be nice, as you say, to also have an option to keep the cross section type of the ports (and ideally, only if it's not "worthwhile" to transition... maybe using a length-based metric)

**Reply by joamatab** [reply]
Awesome, I like the idea of combining some routing parameters into a Class, so we can reduce the amount of parameters passed to the functions

**Comment by sbedi1212** [comment]
Hello, Is there a possibility to get the path length for each manhattan route while using "route_bundle"? I want to calculate the path length difference between two routes, for instance, in an example attached. 
![image](https://github.com/user-attachments/assets/8cf12a15-7c7a-4a83-92c4-b5a2c541b469)

It's important to know the difference in two path lengths (manhattan routes) to accurately determine the time delay for the application I want to build.
I know we can calculate the path length if we use the individual components by using ' .info["length"] ' but what I need is the length of the total manhattan route. Let me know how we can achieve that? Any help is appreciated.

**Reply by joamatab** [reply]
yes, the route has a length in DBU (usually 1nm=1DBU)

**Reply by sbedi1212** [reply]
c = gf.Component('sample_connect') 
mmi1 = c << gf.components.mmi1x2() 
mmi2 = c << gf.components.mmi1x2() 
mmi2.dmove((-80, 40)) 
mmi2.drotate(180) 
routes = gf.routing.route_bundle(c,                         
                                                   [mmi1.ports["o2"], mmi1.ports["o3"]],                         
                                                   [mmi2.ports["o3"], mmi2.ports["o2"]],                         
                                                   radius=6)

for route in routes:
      print(route)

Can you please tell me which function I can use to extract the path length of the routes? Each route is a list and it doesn't show the total length.

**Reply by sebastian-goeldi** [reply]
`route.length` should give you the backbone length in dbu.

**Reply by sbedi1212** [reply]
Thanks for your reply @sebastian-goeldi but the route.length gives only the length_straights and doesn't give the length of the bend-eulers. I was hoping, there could be a direct way to get the total length (including straight waveguides and the bend-eulers). In my opinion, only way to get the total length is through the netlist of the circuit as it gives the length of individual building-blocks of the manhattan route. But this approach could e a bit complex and might involve righting a bit of code. I am working on it at the moment.
If there is still a direct method to achieve it, I am open for any suggestion. Thanks again.

**Reply by joamatab** [reply]
I think this is what you want

```python

import gdsfactory as gf 

c = gf.Component('sample_connect')
mmi1 = c << gf.components.mmi1x2()
mmi2 = c << gf.components.mmi1x2()
mmi2.dmove((-80, 40))
mmi2.drotate(180)
routes = gf.routing.route_bundle(c, [mmi1.ports["o2"],mmi1.ports["o3"]], [mmi2.ports["o3"], mmi2.ports["o2"]],radius=6, cross_section='strip')

for route in routes:
    print(route.length/1e3, 'straight length')
    route_length = 0
    for instance in route.instances:
        route_length += instance.cell.info['length']
    print(route_length, 'total length')
```

65.0 straight length
84.964 total length
65.0 straight length
84.964 total length

**Reply by sbedi1212** [reply]
Thanks a lot @joamatab. This code works. 
In the mean time I figured a way to do it using netlist, although it was a bit complex because the instances name in the route are different then the instances name in the netlist. 
In your code, " instance.cell.info['length'] " was helpful as I was not sure, this can give the length of each instance.
Thanks again!

**Reply by joamatab** [reply]
Thanks for pointing this out,

i'll add this to the routing tutorial docs

![image](https://github.com/user-attachments/assets/52261831-63db-4868-b803-fa0b64209c22)

**Reply by sbedi1212** [reply]
Yes this this sounds good for the tutorial doc.

Also, I would be glad if you can assist me with my query in Q&A section: https://github.com/gdsfactory/gdsfactory/discussions/3378
Thanks @joamatab
